dandischema/metadata.py

```diff
@@ -61,32 +65,89 @@ def generate_context():
+        models.PublishedDandiset.schema_json(indent=2)
     )
     (vdir / "published-asset.json").write_text(
-        models.PublishedAssetMeta.schema_json(indent=2)
+        models.PublishedAsset.schema_json(indent=2)
     )
     (vdir / "context.json").write_text(json.dumps(generate_context(), indent=2))
     return vdir
 
 
-def validate_dandiset_json(data, schema_dir):
+def validate_dandiset_json(data: dict, schema_dir: str) -> None:
     with Path(schema_dir, "dandiset.json").open() as fp:
         schema = json.load(fp)
     jsonschema.validate(data, schema)
 
 
-def validate_asset_json(data, schema_dir):
+def validate_asset_json(data: dict, schema_dir: str) -> None:
     with Path(schema_dir, "asset.json").open() as fp:
         schema = json.load(fp)
     jsonschema.validate(data, schema)
+
+
+def validate(obj, schema_version=None, schema_key=None):
+    """Validate object using pydantic
+
+    Parameters
+    ----------
+    schema_version: str, optional
+       Version of schema to validate against.  If not specified, the schema
+       version specified in `schemaVersion` attribute of object will be used,
+       and if not present - our current DANDI_SCHEMA_VERSION
+    schema_key: str, optional
+        Name of the schema key to be used, if not specified, `schemaKey` of the
+        object will be consulted
+
+     Returns
+     -------
+     None
+
+     Raises
+     --------
+     ValueError:
+        if no schema_key is provided and object doesn't provide schemaKey
+     ValidationError
+        if obj fails validation
+    """
+    schema_key = schema_key or obj.get("schemaKey")
+    if schema_key is None:
+        raise ValueError("Provided object has no known schemaKey")
+    schema_version = schema_version or obj.get("schemaVersion") or DANDI_SCHEMA_VERSION
+    klass = getattr(models, schema_key)
+    klass(**obj)
+
+
+def migrate(obj, to_version=DANDI_SCHEMA_VERSION):
+    if to_version not in ALLOWED_TARGET_SCHEMAS:
+        raise ValueError(f"Current target schemas: {ALLOWED_TARGET_SCHEMAS}.")
+    schema_version = obj.get("schemaVersion")
+    if schema_version == DANDI_SCHEMA_VERSION:
+        return obj
+    if schema_version not in ALLOWED_INPUT_SCHEMAS:
+        raise ValueError(f"Current input schemas supported: {ALLOWED_INPUT_SCHEMAS}.")
+    if version2tuple(schema_version) > version2tuple(to_version):
+        raise ValueError(f"Cannot migrate from {schema_version} to lower {to_version}.")
+    if version2tuple(schema_version) < (0, 3, 2):
+        for contrib in obj["contributor"]:
+            contrib["roleName"] = [
+                val.replace("dandi:", "dcite:") for val in contrib["roleName"]
+            ]
+        for contrib in obj["relatedResource"]:
+            contrib["relation"] = contrib["relation"].replace("dandi:", "dcite:")
+        for access in obj["access"]:
+            access["status"] = "dandi:OpenAccess"
+    obj["schemaVersion"] = to_version
+    return obj
```

Comment:
because validation may still fail. the only time validation passes will be when required fields are in place. migrate can be called on non-publishable metadata.